File: Permutations/Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        for n in nums:
            if len(ans)==0: 
                ans.append([n])
                continue
            N = len(ans)
            for i in range(N):
                l = ans.pop(0)
                for j in range(len(l)+1):
                    new_l = list(l) # 这里效率太低了，先hardcopy再insert很慢，应该直接拼接list
                    new_l.insert(j,n)
                    ans.append(new_l)
        
        if len(ans)==0: return [[]]
        return ans

File: Permutations/Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        if(nums.length==0){
            List<Integer> temp = new ArrayList<Integer>();
            List<List<Integer>> res = new ArrayList<List<Integer>>();
            res.add(temp);
            return res;
        } 
        ArrayList<Integer> list = new ArrayList<Integer>();
        for(int num : nums) list.add(num);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        
        traverse(list, res);
        return res;
    }
    
    public void traverse(ArrayList<Integer> list, List<List<Integer>> res){
        if(list.isEmpty()) return;
        List<Integer> last_per = res.isEmpty()?null:res.get(res.size()-1);
        if(!res.isEmpty()) res.remove(res.size()-1);
        
        for(int i=0; i<list.size(); i++){
            int temp_num = list.get(i);
            list.remove(i);
            // create a deep copy of last_per to new_list
            List<Integer> new_list = new ArrayList<Integer>();
            if(last_per!=null) for(int num:last_per) new_list.add(num);
            // append the temp_num to the last of new_list
            new_list.add(temp_num);
            res.add(new_list);
            // recursion
            traverse(list, res);
            list.add(i, temp_num);
        }
    }
}